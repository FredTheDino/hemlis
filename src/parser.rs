#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    macro_rules! gen_parser {
        ($a:ident, $p:ident) => {
            pub fn $a<'t>(src: &'t str) -> String {
                pub fn inner<'t>(src: &'t str) -> Result<String, String> {
                    use crate::{ast::Ast, grammer, lexer};
                    use std::io::BufWriter;

                    let parser = grammer::$p::new();
                    let out = parser
                        .parse(
                            lexer::lex(&src)
                                .into_iter()
                                .map(|(token, span)| Ok((span.start, token?, span.end))),
                        )
                        .map_err(|x| format!("{:?}", x))?;

                    let mut buf = BufWriter::new(Vec::new());
                    out.show(0, &mut buf).map_err(|x| format!("{:?}", x))?;
                    let inner = buf.into_inner().map_err(|x| format!("{:?}", x))?;
                    String::from_utf8(inner).map_err(|x| format!("{:?}", x))
                }
                match inner(src) {
                    Ok(s) | Err(s) => s,
                }
            }
        };
    }

    gen_parser!(p_header, HeaderParser);
    gen_parser!(p_import, ImportParser);

    #[test]
    fn empty_string() {
        assert_snapshot!(p_header(""));
    }

    #[test]
    fn normal_definition() {
        assert_snapshot!(p_header("module A (a, b, c) where"));
    }

    #[test]
    fn everything_header() {
        assert_snapshot!(p_header(r#"
module A.B.C (a, class B, C, D(..), E(F, G), 
 (+), type (+), module H) where

import A as A
import A.B.C as A.C
import A.B.C (a, class B, 
 C, D(..), E(F, G), (+), type (+))
import A.B.C hiding (foo)
        "#));
    }

    #[test]
    fn simple_import() {
        assert_snapshot!(p_import("import A as A"))
    }

    #[test]
    fn import_with_newlines() {
        assert_snapshot!(p_import("import A (foo\n , bar\n , baz)"))
    }
}
