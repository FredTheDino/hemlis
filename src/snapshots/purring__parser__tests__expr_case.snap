---
source: src/parser.rs
expression: "p_expr(r\"\n        case 1 + 1 of\n            2 -> foo\n            3 | Just _ <- foo bar -> baz\n                    where\n                        baz = 3\n        \")"
---
27 of 27
===
Expr::Case
 9..13
  Expr::Op
   Expr::Number
    Number
     14..15
   QOp
    NULL
    Op
     16..17
   Expr::Number
    Number
     18..19
  CaseBranch
    Binder::Number
     Number
      35..36
   GuardedExpr::Unconditional
    Expr::Ident
     QName
      NULL
      Name
       40..43
  CaseBranch
    Binder::Number
     Number
      56..57
   GuardedExpr::Guarded
     (
       Guard::Binder
        Binder::App
         Binder::Constructor
          QProperName
           NULL
           ProperName
            60..64
         Binder::Wildcard
          65..66
        Expr::App
         Expr::Ident
          QName
           NULL
           Name
            70..73
         Expr::Ident
          QName
           NULL
           Name
            74..77
      Expr::Where
       105..110
       Expr::Ident
        QName
         NULL
         Name
          81..84
        LetBinding::Pattern
         Binder::Var
          Name
           135..138
         Expr::Number
          Number
           141..142
     )
===
